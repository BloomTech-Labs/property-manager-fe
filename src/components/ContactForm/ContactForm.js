/* eslint-disable react/prop-types */
import React from 'react';
import { Formik, Field, Form } from 'formik';
import * as Yup from 'yup';
import { navigate } from '@reach/router';
import { MdEmail, MdMessage, MdPerson } from 'react-icons/md';
import FormErrors from '../LoginForm/FormErrors';
import {
  FormButton,
  ButtonContainer,
  FormHeading,
  I,
  TextInput,
  InputFieldWrapper,
  Label,
  TextAreaInput,
  FormCard
} from '../UI';

const validationSchema = Yup.object().shape({
  name: Yup.string()
    .min(2)
    .max(100)
    .required('Must enter a Name'),
  email: Yup.string()
    .email('Must be a valid email address')
    .max(255, 'Email entered was too long')
    .required('Must enter an Email Address'),
  message: Yup.string()
    .min(8, 'Message must be at least 8 characters')
    .max(255, 'Message is too long')
    .required('Must enter a Message')
});

export default function ContactForm() {
  const encode = data => {
    return Object.keys(data)
      .map(
        // eslint-disable-next-line prefer-template
        key => encodeURIComponent(key) + '=' + encodeURIComponent(data[key])
      )
      .join('&');
  };

  return (
    <FormCard>
      <FormHeading>Contact Us</FormHeading>
      <Formik
        validationSchema={validationSchema}
        initialValues={{
          name: '',
          email: '',
          message: '',
          'bot-field': '',
          'form-name': 'contact'
        }}
        onSubmit={values => {
          fetch('/', {
            method: 'POST',
            headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
            body: encode({
              'form-name': 'contact',
              ...values
            })
          }).then(() => navigate('/thanks'));
        }}
      >
        {({ errors, touched, isSubmitting }) => (
          <Form data-testid="form-element">
            <Field type="hidden" name="bot-field" />
            <Field type="hidden" name="contact" />
            <InputFieldWrapper>
              <Label htmlFor="name">
                <I>
                  <MdPerson />
                </I>{' '}
                Name
              </Label>
              <Field
                as={TextInput}
                placeholder="Enter your name"
                name="name"
                type="text"
              />
              <FormErrors touched={touched.name} message={errors.name} />
            </InputFieldWrapper>
            <InputFieldWrapper>
              <Label htmlFor="email">
                <I>
                  <MdEmail />
                </I>{' '}
                Email
              </Label>
              <Field
                as={TextInput}
                placeholder="Enter your email address"
                name="email"
                type="email"
              />
              <FormErrors touched={touched.email} message={errors.email} />
            </InputFieldWrapper>
            <InputFieldWrapper>
              <Label htmlFor="message">
                <I>
                  <MdMessage />
                </I>{' '}
                Message
              </Label>
              <Field
                as={TextAreaInput}
                placeholder="Enter your message"
                name="message"
                type="textarea"
              />
              <FormErrors touched={touched.message} message={errors.message} />
            </InputFieldWrapper>
            <ButtonContainer>
              <FormButton type="submit" disabled={isSubmitting}>
                Submit
              </FormButton>
            </ButtonContainer>
          </Form>
        )}
      </Formik>
    </FormCard>
  );
}
