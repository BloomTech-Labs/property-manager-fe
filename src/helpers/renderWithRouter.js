/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import { render } from '@testing-library/react';
import {
  createHistory,
  createMemorySource,
  LocationProvider
} from '@reach/router';
import '@testing-library/jest-dom/extend-expect';

export default function renderWithRouter(
  ui,
  { route = '/', history = createHistory(createMemorySource(route)) } = {}
) {
  return {
    ...render(<LocationProvider history={history}>{ui}</LocationProvider>),
    // adding `history` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    history
  };
}
