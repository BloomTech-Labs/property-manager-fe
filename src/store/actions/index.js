// IMPORTS/INITIALIZATION =========================|
// ================================================|
import axios from 'axios';
import axiosAuth from '../../helpers/axiosAuth';
// ------------------------------------------------|
// AUTH ACTIONS ===================================|
// ================================================|
export const AUTH_REQUEST_START = 'AUTH_REQUEST_START';
export const AUTH_REQUEST_SUCCESS = 'AUTH_REQUEST_SUCCESS';
export const AUTH_REQUEST_FAIL = 'AUTH_REQUEST_FAIL';
// ------------------------------------------------|
// LOGIN / SIGNUP ---------------------------------|
export const auth = url => (email, password) => async dispatch => {
  dispatch({ type: AUTH_REQUEST_START });

  try {
    const res = await axios.post(url, {
      email,
      password
    });

    localStorage.setItem('token', res.data.token);

    dispatch({ type: AUTH_REQUEST_SUCCESS, payload: { token: res } });
  } catch (err) {
    console.error(err);
    dispatch({ type: AUTH_REQUEST_FAIL, payload: { errorMessage: err } });
  }
};
// ------------------------------------------------|
// PROPERTY ACTIONS ===============================|
// ================================================|
export const ADD_PROPERTY_START = 'ADD_PROPERTY_START';
export const ADD_PROPERTY_SUCCESS = 'ADD_PROPERTY_SUCCESS';
export const ADD_PROPERTY_FAIL = 'ADD_PROPERTY_FAIL';
export const GET_PROPERTY_START = 'GET_PROPERTY_START';
export const GET_PROPERTY_SUCCESS = 'GET_PROPERTY_SUCCESS';
export const GET_PROPERTY_FAIL = 'GET_PROPERTY_FAIL';
// ------------------------------------------------|
// CREATE A PROPERTY ------------------------------|
export const createProperty = url => property => async dispatch => {
  dispatch({ type: ADD_PROPERTY_START });
  console.log(property);
  try {
    const res = await axios.post(url, { ...property });

    dispatch({
      type: ADD_PROPERTY_SUCCESS,
      payload: {
        created: res
      }
    });
  } catch (err) {
    console.error(err);

    dispatch({ type: ADD_PROPERTY_FAIL, payload: { errorMessage: err } });
  }
};
// ------------------------------------------------|
// GET PROPERTIES ---------------------------------|
export const getProperties = url => async dispatch => {
  dispatch({ type: GET_PROPERTY_START });

  try {
    const res = await axiosAuth().get(url);

    dispatch({
      type: GET_PROPERTY_SUCCESS,
      payload: {
        properties: res.data
      }
    });
  } catch (err) {
    console.error(err);

    dispatch({ type: GET_PROPERTY_FAIL, payload: { errMsg: err } });
  }
};
// ------------------------------------------------|
// USER ACTIONS ===================================|
// ================================================|

// define your user actions here

// ------------------------------------------------|
