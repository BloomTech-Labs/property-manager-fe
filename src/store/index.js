import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import logger from 'redux-logger';
import { getFirebase } from 'react-redux-firebase';
import reducer from './reducers';

const middleware =
  process.ENV === 'production'
    ? [thunk.withExtraArgument(getFirebase)]
    : [thunk.withExtraArgument(getFirebase), logger];

export const makeStore = state =>
  createStore(
    reducer,
    state,
    compose(
      applyMiddleware(...middleware),

      /* eslint-disable */
      window.__REDUX_DEVTOOLS_EXTENSION__
        ? window.__REDUX_DEVTOOLS_EXTENSION__()
        : f => f
    )
  );

export default makeStore();
