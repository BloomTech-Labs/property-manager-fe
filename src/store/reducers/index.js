import { combineReducers } from 'redux';
import authReducer from './authReducer';
import propReducer from './propReducer';
import getUserReducer from './getUserReducer';
import workOrderReducer from './workOrderReducer';

export default combineReducers({
  authReducer,
  propReducer,
  getUserReducer,
  workOrderReducer
});
