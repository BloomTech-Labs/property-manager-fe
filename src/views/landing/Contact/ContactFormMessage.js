/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable react/prop-types */
import React from 'react';
import { navigate } from '@reach/router';
import {
  Button,
  Dialog,
  DialogTitle,
  DialogContent,
  DialogActions,
  Container
} from '@material-ui/core';
import MailSent from '../../../assets/svg/mail-sent.svg';

function ContactFormMessage({ open, handleClose, Transition }) {
  return (
    <>
      <Dialog
        open={open}
        TransitionComponent={Transition}
        keepMounted
        onClose={handleClose}
        aria-labelledby="alert-dialog-slide-title"
        aria-describedby="alert-dialog-slide-description"
      >
        <Container>
          <DialogTitle disableTypography id="alert-dialog-slide-title">
            <h5 className="contactModalHeader">Form Submitted!</h5>
          </DialogTitle>
          <img
            className="contactFormImg"
            src={MailSent}
            alt="Thank you for message"
          />
          <DialogContent>
            <p>Thank you for your feedback!</p>
          </DialogContent>
          <DialogActions>
            <Button autoFocus color="primary" onClick={() => navigate('/')}>
              Close
            </Button>
          </DialogActions>
        </Container>
      </Dialog>
    </>
  );
}

export default ContactFormMessage;
