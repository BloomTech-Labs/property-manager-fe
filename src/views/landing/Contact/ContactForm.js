/* eslint-disable react/prop-types */
/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { Formik, Field, Form } from 'formik';
import { Slide } from '@material-ui/core';
import { MdEmail, MdMessage, MdPerson } from 'react-icons/md';
import FormErrors from '../../../helpers/FormErrors';
import { I } from '../../../components/UI';
import ContactFormMessage from './ContactFormMessage';
import { contactValidation } from './ContactValidation';

const Transition = React.forwardRef(function Transition(props, ref) {
  return <Slide direction="up" ref={ref} {...props} />;
});

export default function ContactForm() {
  // modal logic
  const [open, setOpen] = React.useState(false);

  const handleOpen = () => {
    setOpen(true);
  };

  const handleClose = () => {
    setOpen(false);
  };

  // form logic
  const encode = data => {
    return Object.keys(data)
      .map(
        // eslint-disable-next-line prefer-template
        key => encodeURIComponent(key) + '=' + encodeURIComponent(data[key])
      )
      .join('&');
  };

  return (
    <div className="contact-form-card">
      <h2 className="desktop-form">Contact Us</h2>
      <Formik
        validationSchema={contactValidation}
        initialValues={{
          name: '',
          email: '',
          message: '',
          'bot-field': '',
          'form-name': 'contact'
        }}
        onSubmit={(values, { resetForm }) => {
          fetch('/', {
            method: 'POST',
            headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
            body: encode({
              'form-name': 'contact',
              ...values
            })
          })
            .then(() => {
              resetForm();
              handleOpen();
            })
            // eslint-disable-next-line no-console
            .catch(err => console.error(err));
        }}
      >
        {({ errors, touched, isSubmitting }) => (
          <Form data-testid="form-element">
            <Field type="hidden" name="bot-field" />
            <Field type="hidden" name="contact" />
            <div className="input-wrapper">
              <label htmlFor="name">
                <I>
                  <MdPerson />
                </I>{' '}
                Name
              </label>
              <Field placeholder="Enter your name *" name="name" type="text" />
              <FormErrors touched={touched.name} message={errors.name} />
            </div>
            <div className="input-wrapper">
              <label htmlFor="email">
                <I>
                  <MdEmail />
                </I>{' '}
                Email
              </label>
              <Field
                placeholder="Enter your email address *"
                name="email"
                type="email"
              />
              <FormErrors touched={touched.email} message={errors.email} />
            </div>
            <div className="input-wrapper">
              <label htmlFor="message">
                <I>
                  <MdMessage />
                </I>{' '}
                Message
              </label>
              <Field
                as="textarea"
                rows="4"
                placeholder="Enter your message *"
                name="message"
                type="text"
              />
              <FormErrors touched={touched.message} message={errors.message} />
            </div>
            <div className="submit-btn-wrapper">
              <button
                data-testid="contactBtn"
                className="btn button"
                type="submit"
                disabled={isSubmitting}
              >
                Submit
              </button>
            </div>
          </Form>
        )}
      </Formik>
      <ContactFormMessage
        open={open}
        handleClose={handleClose}
        Transition={Transition}
      />
    </div>
  );
}
