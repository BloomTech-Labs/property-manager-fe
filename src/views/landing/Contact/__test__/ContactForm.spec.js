/* eslint-disable no-undef */
import React from 'react';
import * as rtl from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import store from '../../../../store';
import ContactForm from '../ContactForm';
import renderWithRouter from '../../../../helpers/renderWithRouter';
import ContactFormMessage from '../ContactFormMessage';

afterEach(rtl.cleanup);

test('Checks submit button event for modal', async () => {
  const { container, getByTestId } = await rtl.render(<ContactForm />);

  const appContainer = container;
  expect(await appContainer.innerHTML).toMatch('Contact Us');

  const leftClick = { button: 0 };
  rtl.fireEvent.click(getByTestId(/contactBtn/i), leftClick);
});

test.skip('shows the modal message after submit', async () => {
  const formMessage = 'Form Submitted!';
  const { container, getByText } = await rtl.render(<ContactForm />);
  const {
    modal,
    history: { navigate }
  } = renderWithRouter(<ContactFormMessage>{formMessage}</ContactFormMessage>);

  const leftClick = { button: 0 };
  expect(await container.getByText(formMessage)).toBeNull();
  rtl.fireEvent.click(getByText(/submit/i), leftClick);

  expect(await modal.getByText(formMessage)).toBeInTheDocument();
  await navigate('/');
});

test('calls onSubmit with the fakeUser contact info when submitted', async () => {
  const fakeUser = {
    name: 'George',
    email: 'test@test.com',
    message: 'Hey, can you talk to me?'
  };
  const handleSubmit = jest.fn();
  const { getByPlaceholderText, getByTestId } = await rtl.render(
    <Provider store={store}>
      <ContactForm onSubmit={handleSubmit} />
    </Provider>
  );

  const nameNode = getByPlaceholderText(/enter your name */i);
  const emailNode = getByPlaceholderText(/enter your email address */i);
  const messageNode = getByPlaceholderText(/enter your message */i);
  const formNode = getByTestId('form-element');

  // Act
  rtl.fireEvent.change(nameNode, { target: { value: fakeUser.name } });
  rtl.fireEvent.change(emailNode, { target: { value: fakeUser.email } });
  rtl.fireEvent.change(messageNode, { target: { value: fakeUser.message } });
  await rtl.fireEvent.submit(formNode);

  // Assert
  await rtl.wait(() => {
    expect(handleSubmit).toHaveBeenCalledTimes(0);
  });
});

test('snapshot', () => {
  const { container } = rtl.render(<ContactForm />);
  expect(container.firstChild).toMatchSnapshot();
});
