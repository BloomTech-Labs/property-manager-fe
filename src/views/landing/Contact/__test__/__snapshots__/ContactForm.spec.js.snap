// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  class="contact-form-card"
>
  <h2
    class="desktop-form"
  >
    Contact Us
  </h2>
  <form
    action="#"
    data-testid="form-element"
  >
    <input
      name="bot-field"
      type="hidden"
      value=""
    />
    <input
      name="contact"
      type="hidden"
      value=""
    />
    <div
      class="input-wrapper"
    >
      <label
        for="name"
      >
        <span
          class="css-1bcyk0w"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
            />
          </svg>
        </span>
         
        Name
      </label>
      <input
        name="name"
        placeholder="Enter your name *"
        type="text"
        value=""
      />
    </div>
    <div
      class="input-wrapper"
    >
      <label
        for="email"
      >
        <span
          class="css-1bcyk0w"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V6c0-1.1-.9-2-2-2zm0 4l-8 5-8-5V6l8 5 8-5v2z"
            />
          </svg>
        </span>
         
        Email
      </label>
      <input
        name="email"
        placeholder="Enter your email address *"
        type="email"
        value=""
      />
    </div>
    <div
      class="input-wrapper"
    >
      <label
        for="message"
      >
        <span
          class="css-1bcyk0w"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20 2H4c-1.1 0-1.99.9-1.99 2L2 22l4-4h14c1.1 0 2-.9 2-2V4c0-1.1-.9-2-2-2zm-2 12H6v-2h12v2zm0-3H6V9h12v2zm0-3H6V6h12v2z"
            />
          </svg>
        </span>
         
        Message
      </label>
      <textarea
        name="message"
        placeholder="Enter your message *"
        rows="4"
        type="text"
      />
    </div>
    <div
      class="submit-btn-wrapper"
    >
      <button
        class="btn button"
        data-testid="contactBtn"
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
